app/src/main/java/org/thoughtcrime/securesms/util/BitmapUtil.java

```diff
@@ -0,0 +1,476 @@
+      throws BitmapDecodingException
+  {
+    try {
+      int    quality  = MAX_COMPRESSION_QUALITY;
+      int    attempts = 0;
+      byte[] bytes;
+
+      Bitmap scaledBitmap = GlideApp.with(context.getApplicationContext())
+                                    .asBitmap()
+                                    .load(model)
+                                    .skipMemoryCache(true)
+                                    .diskCacheStrategy(DiskCacheStrategy.NONE)
+                                    .centerInside()
+                                    .submit(maxImageWidth, maxImageHeight)
+                                    .get();
+
+      if (scaledBitmap == null) {
+        throw new BitmapDecodingException("Unable to decode image");
+      }
+
+      Log.i(TAG, String.format(Locale.US,"Initial scaled bitmap has size of %d bytes.", scaledBitmap.getByteCount()));
+      Log.i(TAG, String.format(Locale.US, "Max dimensions %d x %d, %d bytes", maxImageWidth, maxImageHeight, maxImageSize));
+
+      try {
+        do {
+          totalAttempts++;
+          ByteArrayOutputStream baos = new ByteArrayOutputStream();
+          scaledBitmap.compress(format, quality, baos);
+          bytes = baos.toByteArray();
+
+          Log.d(TAG, "iteration with quality " + quality + " size " + bytes.length + " bytes.");
+          if (quality == MIN_COMPRESSION_QUALITY) break;
+
+          int nextQuality = (int)Math.floor(quality * Math.sqrt((double)maxImageSize / bytes.length));
+          if (quality - nextQuality < MIN_COMPRESSION_QUALITY_DECREASE) {
+            nextQuality = quality - MIN_COMPRESSION_QUALITY_DECREASE;
+          }
+          quality = Math.max(nextQuality, MIN_COMPRESSION_QUALITY);
+        }
+        while (bytes.length > maxImageSize && attempts++ < MAX_COMPRESSION_ATTEMPTS);
+
+        if (bytes.length > maxImageSize) {
+          if (sizeAttempt <= MAX_IMAGE_HALF_SCALES) {
+            scaledBitmap.recycle();
+            scaledBitmap = null;
+
+            Log.i(TAG, "Halving dimensions and retrying.");
+            return createScaledBytes(context, model, maxImageWidth / 2, maxImageHeight / 2, maxImageSize, format, sizeAttempt + 1, totalAttempts);
+          } else {
+            throw new BitmapDecodingException("Unable to scale image below " + bytes.length + " bytes.");
+          }
+        }
+
+        if (bytes.length <= 0) {
+          throw new BitmapDecodingException("Decoding failed. Bitmap has a length of " + bytes.length + " bytes.");
+        }
+
+        Log.i(TAG, String.format(Locale.US, "createScaledBytes(%s) -> quality %d, %d attempt(s) over %d sizes.", model.getClass().getName(), quality, totalAttempts, sizeAttempt));
+
+        return new ScaleResult(bytes, scaledBitmap.getWidth(), scaledBitmap.getHeight());
+      } finally {
+        if (scaledBitmap != null) scaledBitmap.recycle();
+      }
+    } catch (InterruptedException | ExecutionException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)